provider/internal/builtin/service_client.go

```diff
@@ -497,3 +473,91 @@ func (b *builtinServiceClient) isFileIncluded(absolutePath string) bool {
 	b.log.V(7).Info("excluding file from search", "file", absolutePath)
 	return false
 }
+
+type walkResult struct {
+	positionParams protocol.TextDocumentPositionParams
+	match          string
+}
+
+func parallelWalk(location string, regex *regexp2.Regexp) ([]walkResult, error) {
+	var positions []walkResult
+	var positionsMu sync.Mutex
+	var eg errgroup.Group
+
+	err := filepath.Walk(location, func(path string, f os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if f.Mode().IsRegular() {
+			eg.Go(func() error {
```

Comment:
We might want to put a limit on this across the engine this could spawn a bunch of go routines I think just setting the limit on the errorgroup.Group should be enough